torch/csrc/jit/codegen/cuda/fusion_segmenter.h

```diff
@@ -0,0 +1,365 @@
+#pragma once
+
+#include <torch/csrc/jit/codegen/cuda/fusion.h>
+#include <torch/csrc/jit/codegen/cuda/ir_base_nodes.h>
+#include <torch/csrc/jit/codegen/cuda/kernel_cache.h>
+#include <torch/csrc/jit/codegen/cuda/scheduler.h>
+#include <torch/csrc/jit/codegen/cuda/scheduler_registry.h>
+
+#include <deque>
+#include <list>
+#include <unordered_set>
+#include <vector>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+//  FusionSegmentFinder
+//    Responsible for going through DAG and proposing things we could try to
+//    fuse together, calls "canGenerateCode" on these proposed segments to see
+//    if they are valid and we can generate code for them.
+//  FusionSegment
+//    A group of exprs that are segmented together
+//  FusionSegmentConnections
+//    Holds vals and what they connect. In other words it's a val that is an
+//    output of a FusionSegment "from" and an input of FusionSegment "to".
+//    There's nothing preventing from a val being between segments twice.
+//    TODO: make sure there's nothing wrong with segmentation on nodes that
+//    have the same value input twice. i.e. (B = A*A)
+
+// Selecting segments to propose is based on the theorem 4.2 in the paper which
+// makes sure when segment the segmented graph will be a DAG (assumes Fusion is
+// already a DAG). The segmentation code relies on assumptions of DAG-ness
+// during segmentation, meaning proposed merging of groups must maintain the DAG
+// property of the graph.
+//
+// Julien Herrmann, Yusuf Özkaya, Bora Uçar, Kamer Kaya, Umit Catalyurek.
+// Multilevel Algorithms for Acyclic Partitioning of Directed Acyclic Graphs.
+// SIAM Journal on Scientific Computing, Society for Industrial and Applied
+// Mathematics, 2019, 41 (4), pp.A2117-A2145. ff10.1137/18M1176865ff.
+// ffhal02306566f
+
+class SegmentedGroup;
+class SegmentCandidateFinder;
+
+// A directed edge on DAG,
+// Wrapper for values, edges between segmented groups which are made up
+// of Exprs. Multiple edges can exist between segmented groups.
+class SegmentedEdge {
+ public:
+  SegmentedEdge(SegmentedGroup* from, SegmentedGroup* to, Val* val)
+      : from_(from), to_(to), val_(val) {}
+  SegmentedGroup* from_;
```

Comment:
if these are public, they should not use the _ suffix (make them private?)

But it may be easier to just make SegmentedEdge a struct instead of a class?

Reply:
Updated and used a struct instead. Thanks!